Framework/Geometry/src/Instrument/ComponentInfo.cpp

```diff
@@ -183,5 +197,136 @@ double ComponentInfo::solidAngle(const size_t componentIndex,
     return shape(componentIndex).solidAngle(relativeObserver, scaleFactor);
   }
 }
+
+/**
+ * Calculates the absolute bounding box for the leaf item at index
+ *
+ * @param index : Component index
+ * @param reference : Optional reference for coordinate system for non-axis
+ *aligned bounding boxes
+ * @return Absolute bounding box.
+ */
+BoundingBox
+ComponentInfo::componentBoundingBox(const size_t index,
+                                    const BoundingBox *reference) const {
+  // Check that we have a valid shape here
+  if (!hasShape(index)) {
+    return BoundingBox(); // Return null bounding box
+  }
+  const auto &s = this->shape(index);
+  BoundingBox absoluteBB = s.getBoundingBox();
+
+  // modify in place for speed
+  const Eigen::Vector3d scaleFactor = m_componentInfo->scaleFactor(index);
+  // Scale
+  absoluteBB.xMin() *= scaleFactor[0];
+  absoluteBB.xMax() *= scaleFactor[0];
+  absoluteBB.yMin() *= scaleFactor[1];
+  absoluteBB.yMax() *= scaleFactor[1];
+  absoluteBB.zMin() *= scaleFactor[2];
+  absoluteBB.zMax() *= scaleFactor[2];
+  // Rotate
+  (this->rotation(index))
+      .rotateBB(absoluteBB.xMin(), absoluteBB.yMin(), absoluteBB.zMin(),
+                absoluteBB.xMax(), absoluteBB.yMax(), absoluteBB.zMax());
+
+  // Shift
+  const Eigen::Vector3d localPos = m_componentInfo->position(index);
+  absoluteBB.xMin() += localPos[0];
+  absoluteBB.xMax() += localPos[0];
+  absoluteBB.yMin() += localPos[1];
+  absoluteBB.yMax() += localPos[1];
+  absoluteBB.zMin() += localPos[2];
+  absoluteBB.zMax() += localPos[2];
+
+  if (reference && !reference->isAxisAligned()) { // copy coordinate system
+
+    std::vector<Kernel::V3D> coordSystem;
+    coordSystem.assign(reference->getCoordSystem().begin(),
+                       reference->getCoordSystem().end());
+
+    // realign to reference coordinate system
+    absoluteBB.realign(&coordSystem);
+  }
+  return absoluteBB;
+}
+
+/**
+ * Compute the bounding box for the component with componentIndex taking into
+ *account
+ * all sub components.
+ *
+ * @param componentIndex : Component index to get the bounding box for
+ * @param reference : Optional reference for coordinate system for non-axis
+ *aligned bounding boxes
+ * @return Absolute bounding box
+ */
+BoundingBox ComponentInfo::boundingBox(const size_t componentIndex,
```

Comment:
Checked this for `StructuredDetector`. Since it is built in exactly the same way as `RectangularDetector` the optimization should be valid 👍